### PR TITLE
fix: recurring events with calendar connections

### LINKED ISSUE #10673: [CAL-2517] Recurring Event - Error while booking a meeting.
### Issue Summary
I get the Couldn't Book Meeting Error while trying to book a recurring meeting. 

### Steps to Reproduce

1. Book a recurring event type

Screenshot

[Chrome logs](https://paste.c-net.org/FelicityComrades)

<sub>[CAL-2517](https://linear.app/calcom/issue/CAL-2517/error-while-booking-a-meeting)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

Fixes that when booking a recurring event with a calendar connection it was always throwing 'no available users found': 
Screenshot 2023-11-06 at 19 30 57

This error was also thrown when user was available on all the recurring dates. 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- Connect Google Calendar -> toggle same calendar to check for conflicts as the one that is set to add the events on
- Create recurring event type 
- Book a slot were you know that all future slot are available too 
- This should pass now

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/lib/parse-dates.ts [hunk 1]
```diff
@@ -98,7 +105,7 @@ export const parseRecurringDates = (
   });
   const dateStrings = times.map((t) => {
     // finally; show in local timeZone again
-    return processDate(t.tz(timeZone), language, { selectedTimeFormat, withDefaultTimeFormat });
+    return processDate(t.tz(timeZone), language, { selectedTimeFormat, withDefaultTimeFormat }, timeZone);
```
Comment (hariombalhara) [inline packages/lib/parse-dates.ts:108]: What is the exact bug that this change fixes?

File: packages/lib/parse-dates.ts [hunk 2]
```diff
@@ -9,15 +9,22 @@ import { parseZone } from "./parse-zone";
 
 type ExtraOptions = { withDefaultTimeFormat?: boolean; selectedTimeFormat?: TimeFormat };
 
-const processDate = (date: string | null | Dayjs, language: string, options?: ExtraOptions) => {
+const processDate = (
+  date: string | null | Dayjs,
+  language: string,
+  options?: ExtraOptions,
+  timeZone?: string
```
Comment (hariombalhara) [inline packages/lib/parse-dates.ts:16]: If we really need timezone to show the correct time here, shouldn't we make it required and fix everywhere it's missing.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 3]
```diff
@@ -408,22 +408,7 @@ async function ensureAvailableUsers(
...
-        // DONE: Decreased computational complexity from O(2^n) to O(n) by refactoring this loop to stop
-        // running at the first unavailable time.
-        let i = 0;
-        while (!foundConflict && i < allBookingDates.length) {
-          foundConflict = checkForConflicts(bufferedBusyTimes, allBookingDates[i++], duration);
-        }
-      } else {
-        foundConflict = checkForConflicts(bufferedBusyTimes, input.dateFrom, duration);
-      }
+      foundConflict = checkForConflicts(bufferedBusyTimes, input.dateFrom, duration);
```
Comment (hariombalhara) [inline packages/features/bookings/lib/handleNewBooking.ts:411]: The change looks good. I was already removing it as part of https://github.com/calcom/cal.com/pull/11774/files

But the interesting thing is that I was thinking that it was dead code but actually it was failing some bookings. 
Did some investigation as to how just enabling a calendar was failing Recurring bookings

- When Calendar is enabled, we don't just fetch the busyTimes for the current slot being booked but instead we fetch till the end of the month. This isn't true for busyTimes fetched from DB. There it's just for the range of the timeslot being booked. The behaviour [changed here 5 months back](https://github.com/calcom/cal.com/pull/9282)
- Due to this, when a single timeslot was checked for conflicts with busytimes, it was actually checking conflicts with busyTime of a recurring timeslot that was just booked in the previous iteration and thus it was always failing.


So, great find @CarinaWolli 👏

**Review by hariombalhara — DISMISSED** [review]
Requesting changes to get clarity on timezone related change, otherwise it's good.

**Comment by keithwillcode** [comment]
Great find, @CarinaWolli. Any tests we could add to make sure this continues to work?

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -408,22 +408,7 @@ async function ensureAvailableUsers(
...
-        eventType.recurringEvent &&
-        recurringDatesInfo?.currentRecurringIndex === 0 &&
-        recurringDatesInfo.allRecurringDates
-      ) {
-        const allBookingDates = recurringDatesInfo.allRecurringDates.map((strDate) => new Date(strDate));
-        // Go through each date for the recurring event and check if each one's availability
-        // DONE: Decreased computational complexity from O(2^n) to O(n) by refactoring this loop to stop
-        // running at the first unavailable time.
-        let i = 0;
-        while (!foundConflict && i < allBookingDates.length) {
```
[the remark this answers is not in the document]
Comment (keithwillcode) [reply packages/features/bookings/lib/handleNewBooking.ts:421]: Wow

**Comment by keithwillcode** [comment]
> > Great find, @CarinaWolli. Any tests we could add to make sure this continues to work?
> 
> I don't think we have a setup yet for testing calendar connections. I think @joeauyeung is working on that. That would be needed to test this

Ok. I thought by looking at the code it would be possible for us to unit test/mock the data this function receives, not necessarily coming from a real calendar connection.

We'll push this PR through to get the bug fixed but let's look into this possibility in a follow-up PR.

**Review by keithwillcode — APPROVED** [review]

**Pull request merged**